pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -259,138 +353,65 @@ private void internalAsyncAddData(T data, AddDataCallback callback, Object ctx){
         if (!batchEnabled) {
             STATE_UPDATER.compareAndSet(this, State.OPEN, State.CLOSED);
             return;
         }
-        // Prevent the reentrant.
+        // If other thread already called "close()", so do nothing.
         if (!STATE_UPDATER.compareAndSet(this, State.OPEN, State.CLOSING)){
-            // Other thread also calling "close()".
             return;
         }
         // Cancel pending tasks and release resources.
         singleThreadExecutorForWrite.execute(() -> {
-            if (state == State.CLOSED){
-                return;
-            }
-            // Failure callback to pending request.
-            // If some request has been flushed, Bookie triggers the callback.
-            failureCallbackByContextAndRecycle(this.flushContext, BUFFERED_WRITER_CLOSED_EXCEPTION);
-            // Cancel task that schedule at fixed rate trig flush.
-            if (timeout == null){
-                log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The field-timeout"
-                        + " is null. managedLedger: " + managedLedger.getName());
-            } else if (timeout.isCancelled()){
-                // TODO How decisions the timer-task has been finished ?
-                this.state = State.CLOSED;
-            } else {
-                if (this.timeout.cancel()) {
-                    this.state = State.CLOSED;
+            try {
+                if (state == State.CLOSED) {
+                    return;
+                }
+                // If some requests are flushed, BK will trigger these callbacks, and the remaining requests in should
+                // fail.
+                failureCallbackByContextAndRecycle(flushContext, BUFFERED_WRITER_CLOSED_EXCEPTION);
+                // Cancel the timing task.
+                if (timeout == null) {
+                    log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The field-timeout"
+                            + " is null. managedLedger: " + managedLedger.getName());
+                } else if (timeout.isCancelled()) {
+                    // TODO How decisions the timer-task has been finished ?
+                    STATE_UPDATER.set(this, State.CLOSED);
                 } else {
-                    // Cancel task failure, The state will stay at CLOSING.
-                    log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The state will"
-                            + " stay at CLOSING. managedLedger: " + managedLedger.getName());
+                    if (this.timeout.cancel()) {
+                        STATE_UPDATER.set(this, State.CLOSED);
+                    } else {
+                        // Cancel task failure, The state will stay at CLOSING.
+                        log.error("Cancel timeout-task that schedule at fixed rate trig flush failure. The state will"
```

Comment:
The only reason `this.timeout.cancel()` will return false is if the state is not INIT.
So either state is CANCELLED, which in this case you're good (you can even check that again inside, or the state is EXPIRED which means the task is going to be executed or is being executed now. In this case, since we're running almost everything in the singleThreadExecutorService, one that timer task will get executed, it will see Timeout is canceled and do nothing.
So just recheck if canceled, if so, close. If not cancelled, close :)

Reply:
I have already rewrite the method `close`:

- Just call `timeout.cancel` and set the state to `closed`, don't care if it's actually executed.
- In method `nextTimingTrigger`, Prevents timing trigger.

These two measures ensure that the Timing trigger will not be executed after `close`.

----

And there are two behaviors that are not as expected( So the above plan is optimal ):

- if the stat is EXPIRED, not means already executed, this just means `expired`(already executed or not). 
- when we call `timeout.cancel`, will close the task asynchrony and return nothing.